Add a check for deprecated references.

gouttegd opened the pull request:
This PR adds a new `check` command intended to perform some additional checks (beyond those already performed by ROBOT) on an ontology.

For now, the only check is a check that no class and no object property is defined relatively to a deprecated entity.

gouttegd committed:
Add a check for deprecated references.

Add a new `check` command intended to perform some additional checks
(beyond those already performed by ROBOT) on an ontology.

For now, the only check is a check that no class and no object property
is defined relatively to a deprecated entity.

gouttegd commented:
@matentzn ⬆️ Follow-up to our discussion this afternoon in the “OBO tech support” meeting, regarding https://github.com/obophenotype/cell-ontology/issues/2967.

matentzn commented on src/main/java/org/incenp/obofoundry/odk/CheckCommand.java:
```diff
@@ -0,0 +1,120 @@
+    private boolean checkDeprecatedReferences(OWLOntology ontology) {
+        boolean pass = true;
+        for ( OWLEntity entity : ontology.getSignature(Imports.INCLUDED) ) {
+            if ( !isInBase(entity.getIRI().toString()) || isDeprecated(ontology, entity) ) {
```
The problem is that we want the check to be for _all entities_ that are _about base entities_..

So if we have

```
CL:123 rdfs:subClassOf :R some UBERON:123
```

And we run this command for CL, we want it to fail if `UBERON:123` is deprecated.

gouttegd replied:
> And we run this command for CL, we want it to fail if UBERON:123 is deprecated.

Yes, and it will – even if you use `--base-iri http://purl.obolibrary.org/obo/CL_` to restrict the check to entities within CL.

Because we only exclude “non-base entities” when iterating over the signature of the ontology, but once an entity has _not_ been excluded, then _all_ the entities that are referenced in its defining axioms (regardless of whether they are “base entities“ or not) are considered.

So in this particular example, when we’ll reach `CL:123`, we’ll find that it is “in base“, then we’ll look to all the entities that it references, find `UBERON:123`, and realises it is deprecated -> boom, error.

(At some point in the process, we will _also_ reach `UBERON:123` in the outer loop, and it will _not_ be considered since it is “off base”. But that has no impact on the check of `CL:123`.)

gouttegd replied:
If you try using it on current CL, it will correctly find that CL:0002038 is referencing GO:0051024, even if you restrict the check to CL entities. The fact that GO:0051024 is not in CL’s base does not matter as long as it is referenced from an entity that is in CL’s base.

matentzn replied:
Got it, thanks!

matentzn commented:
Its impossible how fast of a turnaround this is.. 🪨 🚀

gouttegd commented:
> Its impossible how fast of a turnaround this is.. 

Well, I _have_ to be fast if I want to finish all I have on my plate before I am kicked out of the biocuration field…

matentzn commented on src/main/java/org/incenp/obofoundry/odk/CheckCommand.java:
```diff
@@ -0,0 +1,120 @@
+
+            Set<OWLAxiom> axioms = new HashSet<>();
+            if ( entity instanceof OWLClass ) {
+                axioms.addAll(ontology.getAxioms((OWLClass) entity, Imports.INCLUDED));
```
The implementation of this function is a little lossy (does not get all axioms about the entity). There is this method: https://github.com/ontodev/robot/blob/c77d8628cf1197517668c425adeb5864cb919c22/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L211 called `getAxiomSubjects()` which we use in ROBOT to define what the subject of an axiom is. This is used as the centerpiece to determine if an axiom is included in the base, i.e. "an axiom is `base` if it has at least one subject that is `base`. 

If you wanted the check to be complete with respect to the base specification, it might be best to use that method.

gouttegd replied:
> The implementation of this function is a little lossy (does not get all axioms about the entity)

Depends on what you mean by “about”. It will not return all the axioms where the entity appears at all (that’s what `ontology.getReferencingAxioms()` is for), but that’s not what we need here.

It will return: (A) the _SubClassOf_ axioms where the entity is the subclass, (B) the _EquivalentClasses_ axioms where the entity is one of the operands, and (C) the _DisjointClasses_ axioms where the entity is one of the operands. This is _almost_ all we need.

AFAICT the only missing axioms are the general class axioms, as in `(R some C) SubClassOf X` (where C is an entity we care about) or `(R1 some C1) Disjoint/EquivalentClasses (R2 some C2)` (where either C1 or C2 is an entity we care about). I was planning to deal with them later.

What other axioms could we need to look at here?

> There is this method called `getAxiomSubjects()` which we use in ROBOT to define what the subject of an axiom is

(1) You realize that method is _private_, right? It’s not usable outside of the `RelatedObjectsHelper` class.

(2) You realize that method does the _opposite_ of what is needed here? Given an axiom, it gives you the “subject” entities. Here, we want the axioms, given the entity. Again, this is exactly what `getAxioms(...)` does, modulo the case of the general class axioms.

> If you wanted the check to be complete with respect to the base specification, it might be best to use that method.

That method is _not_ usable, cf (1) above.

And even if it was, what does the “base specification” have anything to do here? We are not trying to determine which axioms belong to the “base” or not.

(Aside: even if the method was usable, I am not sure I’d trust it enough to use it, given that there is a discrepancy between what the Javadoc comment says it does, and what the method does really: the comment says the method returns an empty set if the subject is anonymous, whereas the code clearly returns a set containing all the entities present in the anonymous expression – what is the desired behaviour then? the comment or the code?)

matentzn replied:
> (1) You realize that method is private, right? It’s not usable outside of the RelatedObjectsHelper class.

Oops.

> (2) You realize that method does the opposite of what is needed here? Given an axiom, it gives you the “subject” entities. 

I did realise that, just pointing out that this is the currently best place to figure out the subject of an axiom, which might be needed for reporting on deprecated references and correctly filtering base axioms..

> the comment says the method returns an empty set if the subject is anonymous, whereas the code clearly returns a set containing all the entities present in the anonymous expression

Oops. I hope I will never have to apply to a software engineering position :D 

FAIR enough! I only ever share these thoughts to fill you in what exists, your judgement is far better than mine on such matters!

gouttegd replied:
> I did realise that, just pointing out that this is the currently best place to figure out the subject of an axiom, which might be needed for reporting on deprecated references and correctly filtering base axioms..

Yeah, I guess we could reverse the logic. Instead of:

* looping over entities first,
* excluding off-base entities,
* gathering the axioms about those entities,

we could

* loop over axioms first
* for each axioms collect the subjects, and exclude axioms with an off-base subject,
* get the other non-subject entities and check if they are deprecated.

Seems needlessly complicated in my opinion, especially since we would need to duplicate the logic of the `getAxiomSubjects()` method (since we can't use it directly), whereas for the first approach the OWLAPI already provides what we need (again, modulo the case of the general class axioms, which is trivial to handle separately).

Besides, I disagree with at least one behaviour of the `getAxiomSubjects()` method. Given a `(R1 some C1) EquivalentTo: (R2 some C2)` general class axiom, it will not return any subject at all. That's not a useful behaviour in my opinion. In such a case, both _C1_ and _C2_ can be considered "subjects" - for our purpose here, if any of _C1_ or _C2_ is an in-base entity, then we want to check if any other entity in the axiom is deprecated. The `getAxiomSubjects`-based method would not allow that.

matentzn replied:
>  Given a (R1 some C1) EquivalentTo: (R2 some C2) general class axiom, it will not return any subject at all.

Very very fringe case :P, ok. There should be a complaint here, as every axiom under the sun should belong to _some_ base. https://oboacademy.github.io/obook/reference/base-specification

Thanks for the discussion!

gouttegd replied:
> Very very fringe case

Is it though? In FBbt we have

> '[has part](http://purl.obolibrary.org/obo/BFO_0000051)' **some** ('[has characteristic](http://purl.obolibrary.org/obo/RO_0000053)' **some** [mass](http://purl.obolibrary.org/obo/RO_0000053)) **EquivalentTo** '[has characteristic](http://purl.obolibrary.org/obo/RO_0000053)' **some** [mass](http://purl.obolibrary.org/obo/RO_0000053)

It is apparently coming from CARO, but an annotation on the axiom says it should really belong to BFO – which would make it a pretty common axiom likely to be found in many ontologies.

matentzn replied:
Fair enough; as the spec says, this is an as-yet-undefined case, and you can do whatever you want here!

> IMPORTANT NOTE: Handling of GCIs is still a work in progress. It is not straightforward to define which is the defining entity. For now, unless all entities on the left are in the ontology for which we are making a base, GCIs are never in the base. This is being reviewed and may change in the future.

gouttegd replied:
I think the check for deprecated entity references should have nothing to do with the definition of the base artefact.

If we have axioms like those, and assuming that the _C1_ entity is “in-base”:

> * (R1 **some** C1) **SubClassOf** (R2 **some** C2)
> * (R1 **some** C1) **EquivalentTo** (R2 **some** C2)
> * (R1 **some** C1) **DisjointWith** (R2 **some** C2)

then if _C2_ is obsolete, I think those axioms should be flagged no matter what – regardless of whether we consider that they belong to the base artefact or not.

So the rule I’m planning to use here is: if a general class axiom references _any_ “in-base” entity in its signature (no matter the “position” of the entity, e.g. whether it is in the subclass expression or in the superclass expression in the case of a _SubclassOf_ axiom), then its signature must not reference any deprecated entity (and if it does, we flag it).

In other words, the only case the axioms above should _not_ flagged if _C2_ is obsolete, is if neither _C1_, _C2_, _R1_, and _R2_ are “in-base” entities.


> It is not straightforward to define which is the defining entity

I would argue that it is in fact _impossible_, because it doesn’t make sense. A GCI, almost by definition, has no “defining entity”.

matentzn commented on src/main/java/org/incenp/obofoundry/odk/CheckCommand.java:
```diff
@@ -0,0 +1,120 @@
+            Set<OWLAxiom> axioms = new HashSet<>();
+            if ( entity instanceof OWLClass ) {
+                axioms.addAll(ontology.getAxioms((OWLClass) entity, Imports.INCLUDED));
+            } else if ( entity instanceof OWLObjectProperty ) {
```
OWLDataProperty?

matentzn replied:
OWLNamedIndividual, eg

```
ClassAssertion(:myIndividual, :myFavouriteClass)
```

matentzn replied:
I think if you use the `getAxiomSubjects()` logic above you might get it all.

gouttegd replied:
I was planning to add those cases later on.

`getAxiomSubjects()` would get me nothing for the reasons explained above: it is not usable and it does the opposite of what is needed here.

matentzn reviewed (commented):
My only suggestion is to use our attempt to implement the base spec which is already in ROBOT instead of `ontology.getAxioms((OWLClass) entity, Imports.INCLUDED)`, see comments.

matentzn reviewed (approved):
Excellent! And if some common case is missing we just iterate!

gouttegd committed:
Expand deprecated reference check to more axioms.

When checking for references to deprecated entities, include:

* axioms defining data properties;
* axioms defining individuals;
* general class axioms.

gouttegd committed:
Cache the result of the search for deprecation annotation.

Since checking whether a given entity is deprecated requires iterating
over all the annotations on said entity, and we might have to check the
same entity many times (the entity can be referenced from many different
axioms), we cache the result of that check.
